plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMetadata.java

```diff
@@ -303,13 +313,62 @@ private JdbcTableHandle flushAttributesAsQuery(ConnectorSession session, JdbcTab
         handle = new JdbcTableHandle(
                 handle.getRelationHandle(),
                 handle.getConstraint(),
+                handle.getSortOrder(),
                 OptionalLong.of(limit),
                 handle.getColumns(),
                 handle.getNextSyntheticColumnId());
 
         return Optional.of(new LimitApplicationResult<>(handle, jdbcClient.isLimitGuaranteed(session)));
     }
 
+    @Override
+    public Optional<TopNApplicationResult<ConnectorTableHandle>> applyTopN(
+            ConnectorSession session,
+            ConnectorTableHandle table,
+            long topNCount,
+            List<SortItem> sortItems,
+            Map<String, ColumnHandle> assignments)
+    {
+        if (!isTopNPushdownEnabled(session)) {
+            return Optional.empty();
+        }
+
+        verify(!sortItems.isEmpty(), "sortItems are empty");
+        JdbcTableHandle handle = (JdbcTableHandle) table;
+
+        List<SortItem> resultSortOrder = sortItems.stream()
+                .map(sortItem -> {
+                    verify(assignments.containsKey(sortItem.getName()), "assignments does not contain sortItem: %s", sortItem.getName());
+                    return new SortItem(((JdbcColumnHandle) assignments.get(sortItem.getName())).getColumnName(), sortItem.getSortOrder());
+                })
+                .collect(toImmutableList());
+
+        if (!jdbcClient.supportsTopN(session, handle, assignments, resultSortOrder)) {
+            // JDBC client implementation prevents TopN pushdown for the given table
+            // i.e. when order by on a given type does not match Trino semantic
+            return Optional.empty();
+        }
+
+        if (handle.getSortOrder().isPresent() || handle.getLimit().isPresent()) {
+            if (handle.getLimit().equals(OptionalLong.of(topNCount)) && handle.getSortOrder().equals(Optional.of(resultSortOrder))) {
+                return Optional.of(new TopNApplicationResult<>(handle, jdbcClient.isTopNLimitGuaranteed(session)));
+            }
+
+            // Flush query to apply another sort order or limit
```

Comment:
I don't think this comment adds more information that the encompassing condition. please remove.